source/Halibut/ServiceModel/PendingRequestQueue.cs

```diff
@@ -10,164 +9,96 @@ namespace Halibut.ServiceModel
-        {
-            var pending = await DequeueNextAsync();
-            if (pending == null) return null;
-            return pending.BeginTransfer() ? pending.Request : null;
-        }
-
-        async Task<PendingRequest> DequeueNextAsync()
-        {
-            var first = TakeFirst();
-            if (first != null)
-                return first;
-
-            await Task.WhenAny(hasItems.WaitAsync(), Task.Delay(HalibutLimits.PollingQueueWaitTimeout));
-            hasItems.Reset();
-            return TakeFirst();
-        }
+                    if (task.IsCompleted)
+                    {
+                        if (task.Result.BeginTransfer())
+                        {
+                            return task.Result.Request;
+                        }
+                    }
 
-        PendingRequest TakeFirst()
-        {
-            lock (sync)
-            {
-                if (queue.Count == 0)
                     return null;
-
-                var first = queue[0];
-                queue.RemoveAt(0);
-                return first;
-            }
+                });
         }
-
-        public void ApplyResponse(ResponseMessage response)
-        {
-            if (response == null)
-                return;
-
-            lock (sync)
-            {
-                PendingRequest pending;
-                if (inProgress.TryGetValue(response.Id, out pending))
-                {
-                    pending.SetResponse(response);
-                }
-            }
-        }
-
+        
         class PendingRequest
         {
             readonly RequestMessage request;
             readonly ILog log;
-            readonly ManualResetEventSlim waiter;
-            readonly object sync = new object();
+            readonly AsyncManualResetEvent waiter;
             bool transferBegun;
             bool completed;
 
             public PendingRequest(RequestMessage request, ILog log)
             {
                 this.request = request;
+                this.request.ResponseArrived = ResponseArrived;
                 this.log = log;
-                waiter = new ManualResetEventSlim(false);
+                waiter = new AsyncManualResetEvent(false);
             }
 
-            public RequestMessage Request
+            void ResponseArrived(ResponseMessage responseMessage)
             {
-                get { return request; }
+                SetResponse(responseMessage);
             }
 
-            public void WaitUntilComplete()
+            public RequestMessage Request => request;
+
+            public async Task WaitUntilComplete()
             {
                 log.Write(EventType.MessageExchange, "Request {0} was queued", request);
 
-                var success = waiter.Wait(HalibutLimits.PollingRequestQueueTimeout);
+                var timeoutTask = Task.Delay(HalibutLimits.PollingRequestQueueTimeout);
+                var success = await Task.WhenAny(waiter.WaitAsync(), timeoutTask)
```

Comment:
Not ideal approach. This can leak. For example if the waiter return early the timeout task is still pending. So for each operation you always have two tasks scheduled and one completes only when the timeout is due. I think it would be simpler to ditch the async manual reset event and use TaskCompletionSource<T> with cancellation token source